aten/src/ATen/native/sparse/SparseBlasImpl.cpp

```diff
@@ -2,11 +2,215 @@
+    if (compressed_layout == kSparseCsr) {
+      return t.unsqueeze(-1).unsqueeze_(-1);
+    }
+    else {
+      const auto size_neg_2_blocked = t.size(-2) / blocksize[0];
+      const auto size_neg_1_blocked = t.size(-1) / blocksize[1];
+      auto tiled_sizes = at::DimVector(t.sizes().slice(0, t.dim() - 2));
+      tiled_sizes.push_back(size_neg_2_blocked);
+      tiled_sizes.push_back(blocksize[0]);
+      tiled_sizes.push_back(size_neg_1_blocked);
+      tiled_sizes.push_back(blocksize[1]);
+      return t.reshape(tiled_sizes).transpose(-3, -2);
+    }
+  };
+
+  // Note that sparse values are (..., b0, b1). This means that
+  // the strided input has to be "tilable" to (..., b1, x) with
+  // any x >= 1 such that all the shapes are (block) matrix product
+  // compatible. The matrix product will then have shape (..., b0, x).
+  // This in turn means the the result has to be "tilable" to
+  // (..., b0, x).
+  //
+  // These observations imply the following restrictions:
+  // 1. strided.size(-2) has to be divisible by b1.
+  // 2. result.size(-2) has to be divisible by b0.
+  // 3. both strided.size(-1) and result.size(-1)
+  //    have to be divisible by x.
+  //
+  // Restrictions 1 and 2 are trivially satisfied.
+  // Regarding restriction 3:
+  // it would make sense to take the largest possible x for better
+  // performance since it is very likely that the last dimension
+  // is contiguous. As such, this value is exactly
+  // x = strided.size(-1), since strided.size(-1) == result.size(-1)
+
+  // See the comments above. This is our x.
+  const auto outer_blocksize = n;
+
+  Blocksize strided_blocksize = {blocksize[1], outer_blocksize};
+  const auto strided_tiled = tile_tensor(strided, strided_blocksize);
+
+  // Left argument is (..., b0, b1) and right is (..., b1, x).
+  // This naturally implies the result should be "tilable" as
+  // (..., b0, x).
+  Blocksize result_blocksize = {blocksize[0], outer_blocksize};
+  auto result_tiled = tile_tensor(result, result_blocksize);
+
+  if (compressed_layout == kSparseCsr) {
+    values.unsqueeze_(-1).unsqueeze_(-1);
+  }
+
+  Tensor compressed_indices, plain_indices;
+  std::tie(compressed_indices, plain_indices) = at::sparse_csr::getCompressedPlainIndices(compressed);
+
+  // Select block rows of the strided input that intersect with the block colums of the sparse input.
+  auto strided_tiled_selected_rows = strided_tiled.index_select(-4, plain_indices);
+
+  // Promote to float if output is half or bfloat16 for better precision
+  const auto mm_dtype = (result.scalar_type() == kHalf || result.scalar_type() == kBFloat16)
+    ? kFloat : result.scalar_type();
+  // Now that we know which block rows intersect with which block columns,
+  // we can perform matrix products between pairs of blocks.
+  // NOTE: .to is a no-op when result.scalar_type() == mm_dtype.
+  const auto pairwise_block_mm = values.unsqueeze(-3).to(mm_dtype)
+    .matmul(strided_tiled_selected_rows.to(mm_dtype));
```

Comment:
Well, essentially, we have to reduce `half/bloat16` to `float`, unless we want to have 2x `to(half/bfloat16)` conversions. What are your thoughts on that?

So, again, if mm produces a result in half/bfloat16, we then need to convert it to float, create a buffer of floats, index add into it, and the copy with the to half/bfloat16 conversion.
If, on the other hand, we do a promoted mm, we would then only need a single down-casting in a copy_.

Reply:
Just noting here that we talked through this in our 1:1. I think this is mostly about the dtype we want to use for accumulation. There might be a way to expose a private interface for matmul that allows you to set an output dtype for higher precision accumulation. We can look into this as part of continued performance improvements to this function.